src/common.c

```diff
@@ -186,15 +186,19 @@ psf_log_printf (SF_PRIVATE *psf, const char *format, ...)
 					d = va_arg (ap, int) ;
 
 					if (d < 0)
-					{	d = -d ;
-						sign_char = '-' ;
+					{	sign_char = '-' ;
 						if (lead_char != '0' && left_align == SF_FALSE)
 							width_specifier -- ;
-						} ;
+						
+						u = (d != INT32_MIN) ? -d : INT32_MAX ;
+						}
```

Comment:
another solution would be to cast to unsigned, perform the subtraction, and cast back to signed. While that invokes implementation-defined behaviour, I'm not aware of any platform/compiler that doesn't just take the bit pattern unchanged.

Reply:
I hadn't thought of that. I hadn't thought of that. I'll leave it as it is for now. Let's see if @arthurt has any thoughts on this.

Reply:
Humm, it's also wrong. abs(INT32_MIN) != abs(INT32_MAX) (on two's-compliment hardware, aka, everything.)

Signed-negative-to-unsigned is well defined, as is unsigned unary minus, so doing `-((unsigned)d)` as @SoapGentoo  suggests should be okay.

Reply:
> Humm, it's also wrong. abs(INT32_MIN) != abs(INT32_MAX) (on two's-compliment hardware, aka, everything.)

I don't understand guys, where is UB in line 193? Maybe you didn't notice that `u` is **unsigned**.

Reply:
> Maybe you didn't notice that u is unsigned.

Yeah, but I'm not confident in the order of operations from a plain `u = -d`. Is that equivalent to `d = -d; u = d` or `u = d; u = -u`? We want to avoid the former, use the latter as unsigned overflow is defined.

Force the issue by using `u = -((unsigned)d)`.

Reply:
Okay. Then why not `(unsigned) -d`?  We change sign of signed integer to positive, the value is safe because it is `!= INT32_MIN` branch of ternary operator. Then we cast positive signed integer to unsigned. Looks sane,

Reply:
We, given the predicate of the outer ternary conditional, `(unsigned) -d` is safe, but:

Firstly, it's an unnecessary conditional. The arithmetic required is possible and defined without a conditional. (But this isn't really that important, performance of this function is not of concern.)

Mostly because the second half of the ternary is wrong. `-(INT32_MIN) != INT32_MAX`. (Assuming two's compliment) INT32_MIN = `-2147483648`, INT32_MAX = `214783647`. So, using that conditional, the value -2147483648 would be printed as -2147483647 by this function. (2147483648 is representable as an unsigned int, though, all the way up to 4294967295, aka -1)

You could fix that by doing
```
u = (d != INT32_MIN) ? (unsigned) -d : ((unsigned)INT32_MAX) + 1;
```
or more correctly with
```
u = (d != INT32_MIN) ? (unsigned) -d : -((unsigned)INT32_MIN);
```
at which point it is completely equivalent to:
```
u = -((unsigned)d);
```

Reply:
@arthurt, @SoapGentoo  , I have problems with this code after upgrading to Visual Studio 2022.

here

https://github.com/libsndfile/libsndfile/blob/33e765ccba9a0eb225694fdbf9e299683a8338ee/src/common.c#L193

and here:

https://github.com/libsndfile/libsndfile/blob/33e765ccba9a0eb225694fdbf9e299683a8338ee/src/common.c#L255

With VS2022 on [UWP platform](https://docs.microsoft.com/en-us/windows/uwp/get-started/) [SDL checks](https://docs.microsoft.com/en-us/cpp/build/reference/sdl-enable-additional-security-checks?view=msvc-170) are enabled by default and [C4146 warning](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-2-c4146?view=msvc-170) turned into an error.

Fix this

```c
u = - ((unsigned) d) ; 
```

on

```c
(uint32_t) u = - ((unsigned) d) ; 
```

is correct?